core/src/main/java/io/grpc/PickFirstBalancerFactory2.java

```diff
@@ -0,0 +1,181 @@
+  }
+
+  public static PickFirstBalancerFactory2 getInstance() {
+    return INSTANCE;
+  }
+
+  @Override
+  public LoadBalancer2 newLoadBalancer(LoadBalancer2.Helper helper) {
+    return new PickFirstBalancer(helper);
+  }
+
+  @VisibleForTesting
+  static class PickFirstBalancer extends LoadBalancer2 {
+    private final Helper helper;
+    private Subchannel subchannel;
+
+    @VisibleForTesting
+    static final Attributes.Key<AtomicReference<ConnectivityState>> LAST_STATE =
+        Attributes.Key.of("last-state");
+
+    public PickFirstBalancer(Helper helper) {
+      this.helper = helper;
+    }
+
+    @Override
+    public void handleResolvedAddresses(List<ResolvedServerInfoGroup> servers,
+        Attributes attributes) {
+      // Flatten servers list received from name resolver into single address group. This means that
+      // as far as load balancer is concerned, there's virtually one single server with multiple
+      // addresses so the connection will be created only for the first address (pick first).
+      EquivalentAddressGroup newEag =
+          flattenResolvedServerInfoGroupsIntoEquivalentAddressGroup(servers);
+      if (subchannel == null || !newEag.equals(subchannel.getAddresses())) {
+        if (subchannel != null) {
+          subchannel.shutdown();
+        }
+
+        // NB(lukaszx0): we don't merge `attributes` with `subchannelAttr` because subchannel
+        // doesn't need them. They're describing the resolved server list but we're not taking any
+        // action based on this information.
+        Attributes subchannelAttrs = Attributes.newBuilder()
+            // NB(lukaszx0): because attributes are immutable we can't set new value for the key
+            // after creation but since we can mutate the values we leverge that and set
+            // AtomicReference which will allow mutating state info for given channel.
+            .set(LAST_STATE, new AtomicReference<ConnectivityState>(IDLE))
+            .build();
+
+        subchannel = helper.createSubchannel(newEag, subchannelAttrs);
+        helper.updatePicker(new Picker(PickResult.withSubchannel(subchannel)));
+      }
+    }
+
+    @Override
+    public void handleNameResolutionError(Status error) {
+      helper.updatePicker(new Picker(PickResult.withError(error)));
+    }
+
+    @Override
+    public void handleSubchannelState(Subchannel subchannel, ConnectivityStateInfo stateInfo) {
+      ConnectivityState currentState = stateInfo.getState();
+      AtomicReference<ConnectivityState> lastStateRef = subchannel.getAttributes().get(LAST_STATE);
+      checkState(lastStateRef != null, "subchannel should have LAST_STATE attribute set");
+
+      PickResult pickResult;
+      if (subchannel != this.subchannel || currentState == SHUTDOWN) {
+        return;
+      } else if (currentState == CONNECTING) {
+        pickResult = PickResult.withNoResult();
+      } else if (lastStateRef.get() == TRANSIENT_FAILURE || currentState == READY
```

Comment:
Seems fine. I agree and I'm not sure that there's a lot of benefit in saving few `updatePicker` calls. In any event, since this will be reference impl for future implementor, let's get it right.